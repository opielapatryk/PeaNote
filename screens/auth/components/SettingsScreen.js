import React, { useContext, useEffect, useState } from 'react';
import { Text, View, Button, TextInput, Switch} from 'react-native';
import { AuthContext } from '../../../context/AuthContext';
import { useDispatch, useSelector } from 'react-redux';
import {checkIsAskBeforeStickingNoteFlagOff,deleteAccount,changePassword,askBeforeStick} from '../logic/apiSettingsScreen'
import { styles } from '../../../assets/styles/styles';

const SettingsScreen = () => {
  const { signOut } = useContext(AuthContext);
  const { notes } = useSelector((state) => state.board);
  const dispatchRedux = useDispatch();
  const [message, setMessage] = useState('');
  const [showInput, setShowInput] = useState(true);
  const [oldPassword, setOldPassword] = useState('');
  const [newPassword, setNewPassword] = useState('');
  const [confirmAccountDelete, setConfirmAccountDelete] = useState(false);
  const [askBeforeStickingNoteFlag, setAskBeforeStickingNoteFlag] = useState(false);

  const toggleSwitch = () => askBeforeStick(setAskBeforeStickingNoteFlag,setMessage)

  useEffect(() => {
    checkIsAskBeforeStickingNoteFlagOff(setAskBeforeStickingNoteFlag);
  }, []);


  const renderChangePasswordButtons = () => (
    <>
      <Button title="CHANGE PASSWORD" onPress={handlePasswordChangeButtonPress} />
      <Button title="CONFIRM ACCOUNT DELETE" onPress={()=>deleteAccount(signOut,notes,dispatchRedux)} />
    </>
  );

  const renderDeleteAccountButtons = () => (
    <>
      <Button title="CHANGE PASSWORD" onPress={handlePasswordChangeButtonPress} />
      <Button title="DELETE ACCOUNT" onPress={() => setConfirmAccountDelete(true)} />
    </>
  );

  const handlePasswordChangeButtonPress = () => {
    setConfirmAccountDelete(false);
    setShowInput(false);
    setMessage('');
  };

  const renderDeleteAccountBeforeClickingChangePassword = () => (
    <View>
      <View style={styles.switchRow}>
        <Text style={styles.settingsActionText}>ASK BEFORE STICKING NOTE</Text>
        <Switch
          onValueChange={toggleSwitch}
          value={askBeforeStickingNoteFlag}
        />
      </View>

      {confirmAccountDelete ? renderChangePasswordButtons() : renderDeleteAccountButtons()}

      <Text>{message}</Text>
    </View>
  );

  const renderDeleteAccountAfterClickingChangePassword = () => (
    <View>
      <View style={styles.switchRow}>
        <Text style={styles.settingsActionText}>ASK BEFORE STICKING NOTE</Text>
        <Switch
          onValueChange={toggleSwitch}
          value={askBeforeStickingNoteFlag}
        />
      </View>


      <TextInput placeholder="Old Password" onChangeText={setOldPassword} secureTextEntry />
      <TextInput placeholder="New Password" onChangeText={setNewPassword} secureTextEntry />
      <Button title="CONFIRM NEW PASSWORD" onPress={()=>changePassword(setConfirmAccountDelete,oldPassword,newPassword,setMessage,setShowInput)} />

      {confirmAccountDelete ? (
        <Button title="CONFIRM ACCOUNT DELETE" onPress={()=>deleteAccount(signOut,notes,dispatchRedux)} />
      ) : (
        <Button title="DELETE ACCOUNT" onPress={() => setConfirmAccountDelete(true)} />
      )}

      <Text>{message}</Text>
    </View>
  );

  

  return showInput ? (
    renderDeleteAccountBeforeClickingChangePassword()
  ) : (
    renderDeleteAccountAfterClickingChangePassword()
  );
};

export default SettingsScreen;
