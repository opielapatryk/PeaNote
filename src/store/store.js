import { configureStore } from '@reduxjs/toolkit'
import boardReducer from './notes/boardSlice'
import settingsReducer from './settings/settingsSlice'

export default configureStore({
    reducer:{
        board:boardReducer,
        settings:settingsReducer
    }
})