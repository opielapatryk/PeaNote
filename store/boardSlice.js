import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    notes:[],
    isInput:true,   
    history:[]
}

export const boardSlice = createSlice({
    name:'board',
    initialState,
    reducers:{
        showInput: (state) =>{
            return {...state, isInput: true};
        },
        hideInput: (state) =>{
            return {...state, isInput: false};
        },
        addNote: (state, action) => {
            return {...state, notes:[...state.notes, action.payload],history: [...state.notes]}
        },
        removeNote: (state, action) => {
            return {
              ...state,
              notes: state.notes.filter(note => note.id !== action.payload),
              history: [...state.notes]
            };
          },
          changeInfo: (state, action) => {
            const prevText = state.history.find((note) => note.id === action.payload)?.text || "";
            return {
                ...state,
                notes: state.notes.map((note) => {
                    if (note.id === action.payload) {
                        return {
                            ...note,
                            isInfo: !note.isInfo,
                            text: note.isInfo ? prevText : 'click again to delete note',
                        };
                    }
                    return note;
                }),
                history: [...state.notes]
            };
        }     
    }
})

export const {showInput, addNote, removeNote, hideInput,changeInfo} = boardSlice.actions
export default boardSlice.reducer