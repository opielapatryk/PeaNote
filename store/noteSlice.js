import {createSlice } from '@reduxjs/toolkit'
const initialState = {
    id: 0,
    text:'',
    isNote:true,
    isInfo: false
}
export const noteSlice = createSlice({
    name:'note',
    initialState,
    reducers:{
        showInfo: (state)=>{
            return {...state,isInfo: true,text:'Click againt to delete note'}
        },
        setText: (state,action)=>{
            return {...state,text:action.payload}
        },
        hideInfo: (state)=>{
            return {...state,isInfo: false,text:initialState.text}
        },
        removeNote: (state)=>{
            return {...state,isNote: false}
        },
        isNoteTrue: (state)=>{
            return {...state,isNote: true}
        }
        
    }
})

export const {showInfo, hideInfo,removeNote,setText, isNoteTrue} = noteSlice.actions;

export default noteSlice.reducer