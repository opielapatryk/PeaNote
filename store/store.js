import { configureStore } from '@reduxjs/toolkit'
import boardReducer from './notes/boardSlice'

export default configureStore({
    reducer:{
        board:boardReducer
    }
})