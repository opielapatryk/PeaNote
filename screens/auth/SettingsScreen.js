import { Text, View, Button, TextInput } from 'react-native'
import React, {useContext,useState} from 'react'
import axios from 'axios'
import * as SecureStore from 'expo-secure-store';
import { AuthContext } from '../../context/AuthContext';
import { useDispatch, useSelector} from 'react-redux';
import {removeNote} from '../../store/boardSlice';
// import base64 from 'base-64';

export default SettingsScreen = () => {
  const { signOut } = useContext(AuthContext);
  const { notes } = useSelector((state)=>state.board)
  const dispatch_redux = useDispatch()
  const [message,setMessage] = useState('')
  const [showInput,setShowInput] = useState(true)
  const [old_password,setOldPass] = useState('')
  const [new_password,setNewPass] = useState('')

    async function deleteAccount(){
      try {
        let userID = await SecureStore.getItemAsync('userId');

        const resp = await axios.delete(`http://localhost:8000/api/users/${userID}/`)

        console.log(resp.status);

        if(resp.status === 204){
          signOut()
          notes.forEach(sticker => dispatch_redux(removeNote(sticker.id)))
        }
      } catch (error) {
        console.log(error.message);
      }
    }

    async function changePassword(){
      try {
        const token = await SecureStore.getItemAsync('userToken')

        const resp = await axios.put(`http://localhost:8000/update_password`,{
          "old_password": old_password,
          "new_password": new_password
        },{
          headers:{
            'Content-Type': 'application/json',
            'Authorization': token,
          }
        })

        console.log(resp.status);

        if(resp.status && resp.status === 204){
          setMessage('Password updated!')
        }
        setShowInput(true)
        
      } catch (error) {
        if(error){
          setMessage('Something went wrong! Try provide different password!')
        }
        console.log(error.message);
      }
    }
    
    return (
      showInput?(
      <View>
        <Button title='ASK BEFORE STICKING NOTE'/>
        <Button title='MAKE BOARD PRIVATE'/>
        <Button title='CHANGE PASSWORD' onPress={()=>{
          setShowInput(false)
          setMessage('')
          }}/>
        <Button title='DELETE ACCOUNT' onPress={deleteAccount}/>
        <Text>{message}</Text>
      </View>
      ):(
      <View>
        <Button title='ASK BEFORE STICKING NOTE'/>
        <Button title='MAKE BOARD PRIVATE'/>
        <TextInput placeholder='old password' onChangeText={setOldPass} secureTextEntry/>
        <TextInput placeholder='new password' onChangeText={setNewPass} secureTextEntry/>
        <Button title='CONFIRM NEW PASSWORD' onPress={changePassword}/>
        <Button title='DELETE ACCOUNT' onPress={deleteAccount}/>
        <Text>{message}</Text>
      </View>
      )
    )
}