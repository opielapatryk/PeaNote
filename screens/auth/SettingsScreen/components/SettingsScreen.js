import React, { useEffect, useState } from 'react';
import { View, Text, Switch, TextInput, Pressable,Keyboard,TouchableWithoutFeedback,Image,Dimensions, Linking } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { styles } from '../../../../assets/styles/styles';
import {checkIsAskBeforeStickingNoteFlagOff} from '../functions/checkIsAskBeforeStickingNoteFlagOff';
import {askBeforeStick} from '../functions/askBeforeStick';
import {deleteAccount} from '../functions/deleteAccount';
import {changePassword} from '../functions/changePassword';
import { HANDLE_PASSWORD_CHANGE_BUTTON_PRESS,HANDLE_USERNAME_CHANGE_BUTTON_PRESS } from '../../../constants';
import { setMyimage, setShowInput,setShowInputUsername} from '../../../../store/settings/settingsSlice';
import { setMessage } from '../../../../store/login/loginSlice';
import { useFocusEffect } from '@react-navigation/native';
import { signOutAndClearReduxStore } from '../../Logout/functions/signOutAndClearReduxStore';
import {changeUsername} from '../functions/changeUsername'
import {changeProfilePhoto} from '../functions/changeProfilePhoto'
import * as ImagePicker from 'expo-image-picker'
import auth, { firebase } from '@react-native-firebase/auth';
import { useSafeAreaInsets } from 'react-native-safe-area-context';


const SettingsScreen = () => {
  const [askBeforeStickingNoteFlag, setAskBeforeStickingNoteFlag] = useState(false);
  const [newPassword, setNewPassword] = useState('');
  const [newUsername, setNewUsername] = useState('');
  const [deleteAccountPressed, setDeleteAccountPressed] = useState(false);

  const { notes, pendingNotes } = useSelector((state) => state.board);
  const { showInput, showInputUsername,username,myimage } = useSelector((state) => state.settings);
  const { message } = useSelector((state) => state.login);

  const [image, setImage] = useState(null)

  const dispatch = useDispatch();

  const [status, requestPermission] = ImagePicker.useMediaLibraryPermissions();
  
  useEffect(() => {
    if (image) {
      uploadImage()
      dispatch(setMyimage(image.uri))
      
    }
  }, [image]);

  const pickImage = async () => {
    await requestPermission()

    if(status.granted){
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.All,
        allowsEditing: true,
        aspect: [4,3],
        quality: 1
      });
      const source = {uri: result.assets[0].uri}

      setImage(source)
    }

    if(status.status === ImagePicker.PermissionStatus.DENIED){
      Linking.openSettings()
    }
}; 


const uploadImage = async () => {
  const EMAIL = auth().currentUser.email


  const { uri } = image;

  const path = `${EMAIL}`;

  const reference = firebase.storage().ref(path);

  try {
    const response = await fetch(uri);
    const blob = await response.blob();

    await reference.put(blob);

    setImage(null)
  } catch (error) {
    console.error('Error uploading photo:', error);
  }
} 




  useFocusEffect(
    React.useCallback(() => {
      return ()=>{
        setDeleteAccountPressed(false)
        dispatch(setShowInput(false))
        dispatch(setShowInputUsername(false))
        dispatch(setMessage(''))
      }
    }, [])
  );



  useEffect(() => {
    checkIsAskBeforeStickingNoteFlagOff({ setAskBeforeStickingNoteFlag });
  }, []);

  const handlePasswordChange = () => {
    if (showInput) {
      changePassword({ setDeleteAccountPressed, newPassword, dispatch,setNewPassword });
    } else {
      HANDLE_PASSWORD_CHANGE_BUTTON_PRESS({ setDeleteAccountPressed, dispatch });
    }
  };

  const handleUsernameChange = () => {
    if (showInputUsername) {
      changeUsername({ setDeleteAccountPressed, newUsername, dispatch,setNewUsername });
    } else {
      HANDLE_USERNAME_CHANGE_BUTTON_PRESS({ setDeleteAccountPressed, dispatch });
    }
  };

  const handleDeleteAccount = () => {
    if (deleteAccountPressed) {
      deleteAccount({ notes, dispatch, pendingNotes })
    } else {
      setDeleteAccountPressed(true);
    }
  };
  const insets = useSafeAreaInsets();
  return (
    <TouchableWithoutFeedback 
    onPress={() => Keyboard.dismiss()}>
    <View style={[styles.friendsboard,{
      paddingTop: insets.top,
      flex: 1,
      backgroundColor:'white'
    }]}>
      <View>
      <View style={styles.ProfilePicGrandparent}>
        <View style={styles.ProfilePicParent}>
          {myimage && <Image source={{uri: myimage}} style={styles.ProfilePic}/>}
        </View>
        </View>
        <View style={styles.friendsHeaderRequest}>
          <Text style={styles.friendsHeaderRequestText}>{username}</Text>
        </View>
        <View style={styles.switchRow}>
          <Text style={styles.settingsActionText}>ASK BEFORE STICKING NOTE</Text>
          <Switch
            onValueChange={() => askBeforeStick({ setAskBeforeStickingNoteFlag })}
            value={askBeforeStickingNoteFlag}
          />
        </View>

        

        <Pressable style={styles.friendsHeaderRequest} onPress={async ()=>{
          await pickImage()
        }}>
          <Text style={styles.settingsActionText}>CHANGE PROFILE PHOTO</Text>
        </Pressable>

        {showInput && (
          <TextInput
            style={styles.friendsTextInput}
            placeholder={message?message:"NEW PASSWORD"}
            onChangeText={text=>setNewPassword(text)}
            value={newPassword}
            secureTextEntry
            maxLength={25}
          />   
        )}

        <Pressable style={styles.friendsHeaderRequest} onPress={handlePasswordChange}>
          <Text style={styles.settingsActionText}>{showInput ? 'SET NEW PASSWORD' : 'CHANGE PASSWORD'}</Text>
        </Pressable>

        {showInputUsername && (
          <TextInput
            style={styles.friendsTextInput}
            placeholder={message?message:"USERNAME"}
            onChangeText={text=>setNewUsername(text)}
            value={newUsername}
            maxLength={25}
          />
        )}
        <Pressable style={styles.friendsHeaderRequest} onPress={handleUsernameChange}>
          <Text style={styles.settingsActionText}>{showInputUsername ? 'SET USERNAME' : 'CHANGE USERNAME'}</Text>
        </Pressable>

        <Pressable style={styles.friendsHeaderRequest} onPress={()=>signOutAndClearReduxStore(dispatch)}>
          <Text style={styles.settingsActionText}>LOGOUT</Text>
        </Pressable>
      </View>

      <Pressable style={styles.deleteAccountButton} onPress={handleDeleteAccount}>
        <Text style={styles.deleteAccountText}>
          {deleteAccountPressed ? 'CONFIRM ACCOUNT DELETE' : 'DELETE ACCOUNT'}
        </Text>
      </Pressable>
    </View>
    </TouchableWithoutFeedback>
  );
};

export default SettingsScreen;