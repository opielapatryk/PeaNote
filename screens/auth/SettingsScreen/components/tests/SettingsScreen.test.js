import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';  
import store from '../../../../../store/store'; 
import SettingsScreen from '../SettingsScreen';
import { NavigationContainer } from '@react-navigation/native';
import { act } from 'react-test-renderer';

jest.mock('@react-native-firebase/auth', () => () => {
    return {
        currentUser: {
            email:'test@example.com'
        }
    }
  })

  jest.mock('@react-native-firebase/firestore', () =>()=> {
    return {
      collection: jest.fn(()=>{
        return{
          where: jest.fn(()=>{
            return{
              get: jest.fn(() => Promise.resolve(true))
            }
          }),
        }
      }),
    };
  })


test('Settings screen renders correctly.', async () => {
  const tree = renderer.create(
    <Provider store={store}>
      <NavigationContainer>
        <SettingsScreen/>
      </NavigationContainer>
    </Provider>
  ).toJSON();

  await act(async () => {
    expect(tree).toMatchSnapshot();
  })
});
