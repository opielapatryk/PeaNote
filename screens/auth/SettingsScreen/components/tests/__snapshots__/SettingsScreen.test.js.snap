// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pending note renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFBE4",
      "flex": 1,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#FFF3B2",
        "borderRadius": 20,
        "borderWidth": 1,
        "elevation": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 20,
        "padding": 3,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.18,
        "shadowRadius": 1,
        "width": 300,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 12,
          "letterSpacing": 0.5,
          "paddingRight": 10,
        }
      }
    >
      ASK BEFORE STICKING NOTE
    </Text>
    <RCTSwitch
      accessibilityRole="switch"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 31,
          "width": 51,
        }
      }
      value={false}
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#FFF3B2",
        "borderRadius": 20,
        "borderWidth": 1,
        "elevation": 1,
        "marginBottom": 20,
        "padding": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.18,
        "shadowRadius": 1,
        "width": 300,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 12,
          "letterSpacing": 0.5,
        }
      }
    >
      CHANGE PASSWORD
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#FFF3B2",
        "borderRadius": 20,
        "borderWidth": 1,
        "elevation": 1,
        "marginBottom": 20,
        "padding": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.18,
        "shadowRadius": 1,
        "width": 300,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 12,
          "letterSpacing": 0.5,
        }
      }
    >
      DELETE ACCOUNT
    </Text>
  </View>
  <Text
    style={
      {
        "alignSelf": "center",
      }
    }
  />
</View>
`;
